### DISCUSSION TITLE
Open Avalonia window on a new thread

### DISCUSSION THREAD

**Discussion by RaoHammas** [AvaloniaUI/Avalonia#16716] [Q&A] [answered]
Hi,

I'm trying to open Avalonia window on a new thread but it's not working. It throws `Call from invalid thread` error when window is created. Similar code works fine in WPF application. Is it not allowed to open Window on another thread in Avalonia?

    public void Show<TViewModel>(bool newThread = true) where TViewModel : class
    {
        if (newThread)
        {
            var thread = new Thread(() =>
            {
                var window = CreateAndSetupWindow<TViewModel>();
                window.Closed += (s, e) => Dispatcher.UIThread.ExitAllFrames();

                Dispatcher.UIThread.InvokeAsync(() =>
                {
                    // Show the window on the main UI thread
                    window.Show();
                });

                Dispatcher.UIThread.MainLoop(CancellationToken.None);
            });

            thread.SetApartmentState(ApartmentState.STA);
            thread.Start();
        }
        else
        {
            var window = CreateAndSetupWindow<TViewModel>();
            window.Show();
        }
    }

    private Window CreateAndSetupWindow<TViewModel>() where TViewModel : class
    {

        var windowTypeName = typeof(TViewModel).Name.Replace("ViewModel", "");

        //This 👇 line throws exception. Even if i do something like var window = new SomeWindow();
        
        var window = _serviceProvider.GetRequiredKeyedService<Window>(windowTypeName); 

        .....
        .....
        return window!;
    }

**Comment by stevensrmiller** [comment, marked as answer]
The short answer is, "No, you can't open an Avalonia window on another thread." However, it is pretty easy to have worker threads submit tasks to run on the Avalonia UI thread. Look at [this page](https://docs.avaloniaui.net/docs/guides/development-guides/accessing-the-ui-thread#dispatcher-priority) from the Avalonia documentation. And [this sample code](https://github.com/stevensrmiller/AvaloniaWithoutXAML/tree/main/samples/MultiThreading) I wrote might interest you too. Good luck!

**Reply by RaoHammas** [reply]
Thanks for the reply. Invoking main (and only) GUI thread is not feasible in many cases. If one window has rapid updates happening or if it is stuck for whatever reason, it will hang/freeze the whole application. In WPF, we handled this scenario by opening each window on a separate thread.

**Reply by stevensrmiller** [reply]
I'm fairly new to Avalonia and don't know much at all about WPF. I do know a lot about Java/Swing, which also requires all UI API calls to be made from one thread (Swing calls this the "Event Dispatcher Thread," or "EDT"). The problems you are describing are easily avoided in the Swing system by simply queueing the UI calls to the EDT. What I see of Avalonia so far tells me it will work just as well to use its system of posting UI calls. What kind of problem are you working on where you don't see this as feasible?

**Reply by RaoHammas** [reply]
I'm testing Avalonia for a high-performance trading app. Where latency will be close to zero. So, huge data and updating rapidly on multiple windows can create problem. For the same system we use multithreaded windows in WPF, and it works great.

**Reply by stevensrmiller** [reply]
Well, try it before you worry about it. Remember that the rendering is all done on the UI thread. The only overhead you're going to encounter is in thread context switching. You'd have that anyway in a multi-threaded system that had multiple rendering threads.

**Reply by RaoHammas** [reply]
This is what i wanted from avalonia but it seems you are right and it's not possible.
[MSDN#multiple-windows-multiple-In-WPF](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/threading-model?view=netframeworkdesktop-4.8&redirectedfrom=MSDN#multiple-windows-multiple-threads:~:text=this%20situation%20next.-,Multiple%20windows%2C%20multiple%20threads,-Some%20WPF%20applications)

**Reply by stevensrmiller** [reply]
For a trading app (I've written a number of those, actually), I can't imagine why it would be necessary. Consider what it says in the page you linked to:

> Some WPF applications require multiple top-level windows. It's perfectly acceptable for one Thread/Dispatcher combination to manage multiple windows, but sometimes several threads do a better job. This is especially true if there's any chance that one of the windows will monopolize the thread.

To keep any GUI responsive, you have to insure that nothing will, as they put it, "monopolize" the thread. You use a worker thread to avoid that. Whenever the worker would update its window, you'll have the worker post an update operation to the UI thread. Then the worker can go on with whatever it was doing, and the UI thread will remain responsive.

Now also look at the next paragraph on your WPF page:

> Windows Explorer works in this fashion. Each new Explorer window belongs to the original process, but it's created under the control of an independent thread. When Explorer becomes nonresponsive, such as when looking for network resources, other Explorer windows continue to be responsive and usable.

Again, that makes sense because any one of those Explorer windows could block on something that shouldn't block any of the others. If your system is going to be real-time, that should never happen. But, even if it does, you can avoid the problem entirely by having potentially blocking operations happen on worker threads, and all UI activity on the UI thread. The UI thread will never block. And any blocking you might experience on another thread would not go away simply because that thread could also do its own UI operations.

As they say about WPF, Avalonia has no problem with any number of "main" windows all being managed by one UI thread. My repo has two examples showing how you can do this, [this one](https://github.com/stevensrmiller/AvaloniaWithoutXAML/tree/main/samples/MultiWindowNoMain), and [this other one](https://github.com/stevensrmiller/AvaloniaWithoutXAML/tree/main/samples/MultiWindowCancelToken). There are even more approaches one could use. But I really wouldn't worry about any noticeable overhead arising from posting UI calls from a worker thread to the UI thread. Heck, in a trading app, you're going to be looking at delays of several milliseconds, at least, just in network latency. The thread-switch overhead is going to be in _microseconds_. You just won't ever notice it.

**Comment by stevensrmiller** [comment]
Putting this in as an answer to your original question, so you can mark it closed if you want. Yes, you are correct that it is not possible to open a window on a thread other than the UI thread in Avalonia.

**Comment by kekekeks** [comment]
It's a platform limitation for macOS, WASM, Android and iOS. We can't have more than one UI thread there. 

We could potentially implement multiple UI threads for Windows and Linux at some point, but right now there is not enough demand (no demand at all, actually) from our paying customers to justify an effort like that.
